src/ImGui.NET.SampleProgram.FNA/ImGuiRenderer.cs

```diff
@@ -0,0 +1,346 @@
+                _vertexBuffer?.Dispose();
+
+                _vertexBufferSize = (int)(drawData->TotalVtxCount * 1.5f);
+                _vertexBuffer = new VertexBuffer(_graphicsDevice, DrawVertDeclaration.Declaration, _vertexBufferSize, BufferUsage.None);
+            }
+
+            if (drawData->TotalIdxCount > _indexBufferSize)
+            {
+                _indexBuffer?.Dispose();
+
+                _indexBufferSize = (int)(drawData->TotalIdxCount * 1.5f);
+                _indexBuffer = new IndexBuffer(_graphicsDevice, IndexElementSize.SixteenBits, _indexBufferSize, BufferUsage.None);
+            }
+
+            // Update buffers with the latest data from ImGui
+            int vtxOffset = 0;
+            int idxOffset = 0;
+
+            for (int n = 0; n < drawData->CmdListsCount; n++)
+            {
+                var cmdList = drawData->CmdLists[n];
+
+                _vertexBuffer.SetDataPointerEXT(vtxOffset * DrawVertDeclaration.Size, (IntPtr)cmdList->VtxBuffer.Data, cmdList->VtxBuffer.Size * DrawVertDeclaration.Size, SetDataOptions.None);
+                _indexBuffer.SetDataPointerEXT(idxOffset * (int)sizeof(ushort), (IntPtr)cmdList->IdxBuffer.Data, cmdList->IdxBuffer.Size * sizeof(ushort), SetDataOptions.None);
+
+                vtxOffset += cmdList->VtxBuffer.Size;
+                idxOffset += cmdList->IdxBuffer.Size;
+            }
+        }
+
+        private unsafe void RenderCommandLists(DrawData* drawData)
+        {
+            _graphicsDevice.SetVertexBuffer(_vertexBuffer);
+            _graphicsDevice.Indices = _indexBuffer;
+
+            int vtxOffset = 0;
+            int idxOffset = 0;
+
+            for (int n = 0; n < drawData->CmdListsCount; n++)
+            {
+                var cmdList = drawData->CmdLists[n];
+
+                for (int cmdi = 0; cmdi < cmdList->CmdBuffer.Size; cmdi++)
+                {
+                    var drawCmd = &(((DrawCmd*)cmdList->CmdBuffer.Data)[cmdi]);
+
+                    if (!_loadedTextures.ContainsKey(drawCmd->TextureId)) throw new InvalidOperationException($"Could not find a texture with id '{drawCmd->TextureId}', please check your bindings");
+
+                    _graphicsDevice.ScissorRectangle = new Rectangle(
+                        (int)drawCmd->ClipRect.X,
+                        (int)drawCmd->ClipRect.Y,
+                        (int)(drawCmd->ClipRect.Z - drawCmd->ClipRect.X),
+                        (int)(drawCmd->ClipRect.W - drawCmd->ClipRect.Y)
+                    );
+
+                    var effect = UpdateEffect(_loadedTextures[drawCmd->TextureId]);
+
+                    foreach (var pass in effect.CurrentTechnique.Passes)
+                    {
+                        pass.Apply();
+
+                        _graphicsDevice.DrawIndexedPrimitives(
+                            primitiveType: PrimitiveType.TriangleList,
+                            baseVertex: vtxOffset,
+                            minVertexIndex: vtxOffset,
```

Comment:
I don't really know what this "minVertexIndex" parameter does. Is this right? It feels weird that it's the same as "baseVertex".

Reply:
Apparently the "baseVertex" is the actual offset in the vertex array, from which all other work is done. "minVertexIndex" can then be provided as an additional offset, relative to "baseVertex". I haven't found out why though.

The render seems to still work, even when setting the "minVertexIndex" to 0 or 100000. FNA's source describes the parameter as a "hint", and the OpenGL doc of the underlying function mentions a mechanism in which overflowing vertices are still drawn.

I've changed the minVertexIndex to 0, which is the most "correct" value I can think of. I imagine that some of these parameters may be lost in translation a bit, considering they came from a Direct3D backend.

Reply:
Interesting. I tried to look them up as well and, to the best that I could find, they are just "hints" for some legacy renderer or something like that. It seems like neither of them really matter and aren't relevant for modern API's.